domain/application/service/application.go

```diff
@@ -1245,6 +1245,71 @@ func (s *Service) GetApplicationDisplayStatus(ctx context.Context, appID coreapp
 	return derivedApplicationStatus, nil
 }
 
+// GetApplicationAndUnitStatusesForUnitWithLeader returns the display status
+// of the application the specified unit belongs to, and the workload statuses
+// of all the units that belong to that application, indexed by unit name.
+// If no application is found for the unit name, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned
+func (s *Service) GetApplicationAndUnitStatusesForUnitWithLeader(
+	ctx context.Context,
+	unitName coreunit.Name,
+) (
+	*corestatus.StatusInfo,
+	map[coreunit.Name]corestatus.StatusInfo,
+	error,
+) {
+	if err := unitName.Validate(); err != nil {
+		return nil, nil, internalerrors.Errorf("unit name: %w", err)
+	}
+
+	appName := unitName.Application()
+
+	var applicationDisplayStatus *corestatus.StatusInfo
+	var unitWorkloadStatuses map[coreunit.Name]corestatus.StatusInfo
+
+	err := s.leaderEnsurer.WithLeader(ctx, appName, unitName.String(), func(ctx context.Context) error {
+		appID, err := s.st.GetApplicationIDByName(ctx, appName)
+		if err != nil {
+			return internalerrors.Errorf("getting application id: %w", err)
+		}
+		applicationStatus, err := s.st.GetApplicationStatus(ctx, appID)
+		if err != nil {
+			return internalerrors.Errorf("getting application status: %w", err)
+		}
+		workloadStatuses, cloudContainerStatuses, err := s.st.GetUnitWorkloadAndCloudContainerStatusesForApplication(ctx, appID)
+		if err != nil {
+			return internalerrors.Errorf("getting unit workload and container statuses")
+		}
+
+		unitWorkloadStatuses, err = decodeUnitWorkloadStatuses(workloadStatuses)
+		if err != nil {
+			return internalerrors.Errorf("decoding unit workload statuses: %w", err)
+		}
+
+		if applicationStatus.Status != application.WorkloadStatusUnset {
+			applicationDisplayStatus, err = decodeWorkloadStatus(&application.UnitStatusInfo[application.WorkloadStatusType]{
+				StatusInfo: *applicationStatus,
+				Present:    true,
+			})
```

Comment:
I'd rather have a function that turned a applicationStatus into a workload status, or have `decodeUnitWorkloadStatus` vs `decodeApplicationWorkloadStatus`. They've clearly diverged on decoding now.

Reply:
Yeah good point. I'll fix this everywhere it is currently used